File: Manual_5/problem_1.c
#include <stdio.h>

void priority_scheduling(int processes[], int n, int priorities[]) {
    int waiting_time[n], turnaround_time[n];
    int total_waiting_time = 0, total_turnaround_time = 0;

    waiting_time[0] = 0;

    // Calculate waiting time for each process
    for (int i = 1; i < n; i++) {
        waiting_time[i] = waiting_time[i - 1] + processes[i - 1];
    }

    // Calculate turnaround time for each process
    for (int i = 0; i < n; i++) {
        turnaround_time[i] = waiting_time[i] + processes[i];
        total_waiting_time += waiting_time[i];
        total_turnaround_time += turnaround_time[i];
        printf("Process %d\t Priority %d\t Waiting Time %d\t Turnaround Time %d\n", i + 1, priorities[i], waiting_time[i], turnaround_time[i]);
    }

    printf("\nAverage Waiting Time: %.2f\n", (float)total_waiting_time / n);
    printf("Average Turnaround Time: %.2f\n", (float)total_turnaround_time / n);
}

int main() {
    int processes[] = {5, 3, 8, 6, 2};
    int priorities[] = {2, 1, 3, 2, 4};
    int n = sizeof(processes) / sizeof(processes[0]);

    printf("Priority CPU Scheduling Algorithm:\n");
    priority_scheduling(processes, n, priorities);

    return 0;
}

File: Manual_5/problem_1.py
#Priority CPU Scheduling Algorithm:
def priority_scheduling(processes, priorities):
    n = len(processes)
    waiting_time = [0] * n
    turnaround_time = [0] * n

    # Sort processes based on priority
    sorted_processes = sorted(zip(processes, priorities), key=lambda x: x[1])

    waiting_time[0] = 0

    # Calculate waiting time for each process
    for i in range(1, n):
        waiting_time[i] = waiting_time[i - 1] + sorted_processes[i - 1][0]

    # Calculate turnaround time for each process
    for i in range(n):
        turnaround_time[i] = waiting_time[i] + sorted_processes[i][0]

    print("Process\tPriority\tWaiting Time\tTurnaround Time")
    total_waiting_time = 0
    total_turnaround_time = 0
    for i in range(n):
        total_waiting_time += waiting_time[i]
        total_turnaround_time += turnaround_time[i]
        print(f"{sorted_processes[i][0]}\t{sorted_processes[i][1]}\t\t{waiting_time[i]}\t\t{turnaround_time[i]}")

    print(f"\nAverage Waiting Time: {total_waiting_time / n}")
    print(f"Average Turnaround Time: {total_turnaround_time / n}")


# Example usage:
processes = [1, 2, 3, 4, 5]
priorities = [2, 1, 3, 2, 4]
priority_scheduling(processes, priorities)
